gateway/ln-gateway/src/db.rs

```diff
@@ -508,6 +523,77 @@ async fn migrate_to_v4(mut ctx: MigrationContext<'_>) -> Result<(), anyhow::Erro
     Ok(())
 }
 
+async fn migrate_to_v5(mut ctx: MigrationContext<'_>) -> Result<(), anyhow::Error> {
+    let mut dbtx = ctx.dbtx();
+
+    // We need to migrate all isolated database entries to be behind the 0x10
+    // prefix. The problem is, if there is a `FederationId` that starts with
+    // 0x04, we cannot read the `FederationId` because the database will be confused
+    // between the isolated DB and the `FederationIdKey` record. To solve this,
+    // what we do first to try and see if there are any entries that begin with
+    // 0x0404. This indicates a joined federation that will have a
+    // problem decoding the ID.
+    let problem_fed_configs = dbtx
+        .raw_find_by_prefix(&[0x04, 0x04])
+        .await?
+        .collect::<BTreeMap<_, _>>()
+        .await;
+    for (problem_key, problem_fed_config) in problem_fed_configs {
+        let federation_id = FederationId::consensus_decode_vec(
+            problem_key[1..33].to_vec(),
+            &ModuleDecoderRegistry::default(),
+        )?;
+        tracing::warn!(
+            ?federation_id,
+            "Found a FederationConfig entry that will cause issues decoding"
+        );
+        let federation_id_bytes = problem_key[1..3].to_vec();
+
+        // Remove the `FederationConfig` entry so that the migration of the isolated
+        // database doesn't accidentally migrate that too.
+        dbtx.raw_remove_entry(&problem_key).await?;
+
+        let isolated_entries = dbtx
+            .raw_find_by_prefix(&federation_id_bytes)
+            .await?
+            .collect::<BTreeMap<_, _>>()
+            .await;
+        for (mut key, value) in isolated_entries {
+            let mut new_key = vec![DbKeyPrefix::ClientDatabase as u8];
+            new_key.append(&mut key);
+            dbtx.raw_remove_entry(&key).await?;
+            dbtx.raw_insert_bytes(&new_key, &value).await?;
+        }
```

Comment:
This chunk of code is almost identical to the one below, other than it being collected into a `BTreeMap` instead of a `Vec`. Is there a reason for that difference? If not, could we define a helper function at the top of `migrate_to_v5()`? Something like this:

```rust
    async fn migrate_client_entries(
        dbtx: &mut DatabaseTransaction<'_>,
        prefix: &[u8],
    ) -> Result<(), anyhow::Error> {
        let isolated_entries = dbtx
            .raw_find_by_prefix(prefix)
            .await?
            .collect::<BTreeMap<_, _>>()
            .await;

        for (mut key, value) in isolated_entries {
            let mut new_key = vec![DbKeyPrefix::ClientDatabase as u8];
            new_key.append(&mut key);
            dbtx.raw_remove_entry(&key).await?;
            dbtx.raw_insert_bytes(&new_key, &value).await?;
        }

        Ok(())
    }
```